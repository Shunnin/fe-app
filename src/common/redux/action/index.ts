import { trimStart, isUndefined } from 'lodash-es';

const DELIMITER = '/';
const REQUEST = 'REQUEST';
const SUCCESS = 'SUCCESS';
const ERROR = 'ERROR';

interface IAsyncActionType {
  REQUEST: string;
  SUCCESS: string;
  ERROR: string;
}

interface IAsyncActionCreator {
  request: Function;
  success: Function;
  error: Function;
}

export const createActionType = (type: string, module: string = '') => {
  return trimStart(`${module}${DELIMITER}${type}`, DELIMITER);
};

export const createAsyncActionType = (type: string, module: string = ''): IAsyncActionType => {
  return [REQUEST, SUCCESS, ERROR].reduce(
    (acc, asyncType) => {
      const actionType = `${type}_${asyncType}`;

      return Object.assign(acc, {
        [asyncType]: createActionType(actionType, module),
      });
    },
    {
      REQUEST: '',
      SUCCESS: '',
      ERROR: '',
    },
  );
};

export const createActionCreator = (type: string, ...argNames: string[]) => {
  return (...args) => {
    const action = { type };

    argNames.forEach((arg, index) => {
      if (!isUndefined(args[index])) {
        action[argNames[index]] = args[index];
      }
    });

    return action;
  };
};

export const createAsyncActionCreator = (asyncActionType: IAsyncActionType): IAsyncActionCreator => ({
  request: createActionCreator(asyncActionType.REQUEST, 'payload'),
  success: createActionCreator(asyncActionType.SUCCESS, 'response', 'payload'),
  error: createActionCreator(asyncActionType.ERROR, 'error', 'payload'),
});
