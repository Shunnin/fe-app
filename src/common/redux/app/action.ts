import { CHANGE_TEMP_UNIT } from './action-type';
import { IChangeTempUnit } from './type';

export const changeTempUnit = (payload: boolean): IChangeTempUnit => ({
  type: CHANGE_TEMP_UNIT,
  payload,
});
