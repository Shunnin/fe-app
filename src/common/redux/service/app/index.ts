export { CHANGE_TEMP_UNIT, changeTempUnit } from './app.action';
export { appReducer } from './app.reducer';
export { APP_MODULE } from './app.constant';
