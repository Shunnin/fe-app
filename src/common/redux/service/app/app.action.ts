import { createActionType, createActionCreator } from '../../action';

import { APP_MODULE } from './app.constant';

export const CHANGE_TEMP_UNIT = createActionType('CHANGE_TEMP_UNIT', APP_MODULE);

export const changeTempUnit = createActionCreator(CHANGE_TEMP_UNIT, 'payload');
