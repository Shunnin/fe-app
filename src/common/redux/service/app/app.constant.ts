import { IDefaultAction } from '../../reducer';

import { CHANGE_TEMP_UNIT } from './app.action';

export const APP_MODULE = 'appModule';

export type IChangeTempUnit = {
  type: typeof CHANGE_TEMP_UNIT;
  payload: boolean;
};

export interface IAppState {
  tempUnit: string;
}

export type AppActions = IChangeTempUnit | IDefaultAction;
