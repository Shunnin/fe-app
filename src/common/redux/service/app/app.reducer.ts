import { TEMP_UNIT, TEMP_UNIT_TOGGLE_VALUE } from '../../../utility';

import { CHANGE_TEMP_UNIT } from './app.action';
import { AppActions, IAppState } from './app.constant';

const initialState: IAppState = {
  tempUnit: TEMP_UNIT.CELSIUS,
};

const handleChangeTempUnit = (state: IAppState, payload: boolean) => {
  const tempUnit = TEMP_UNIT_TOGGLE_VALUE[payload];

  return { ...state, tempUnit };
};

export const appReducer = (state = initialState, action: AppActions) => {
  const { type, payload } = action;

  switch (type) {
    case CHANGE_TEMP_UNIT:
      return handleChangeTempUnit(state, payload);

    default:
      return { ...state };
  }
};
