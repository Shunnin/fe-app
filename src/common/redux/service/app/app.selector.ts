import { get } from 'lodash-es';

import { createSelector, combineSelectors } from '../../selector';

import { APP_MODULE } from './app.constant';

interface IState {
  [APP_MODULE]?: {
    tempUnit: string;
  };
}

const appStateSelector = (state: IState) => state[APP_MODULE];

export const tempUnitSelector = createSelector(appStateSelector, moduleState => {
  return get(moduleState, 'tempUnit');
});

export const appSelectors = combineSelectors({
  tempUnit: tempUnitSelector,
});
