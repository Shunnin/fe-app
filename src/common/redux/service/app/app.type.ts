import { CHANGE_TEMP_UNIT } from './app.action-type';

export type IChangeTempUnit = {
  type: typeof CHANGE_TEMP_UNIT;
  payload: boolean;
};

export interface IAppState {
  tempUnit: string;
}

export type AppActions = IChangeTempUnit;
