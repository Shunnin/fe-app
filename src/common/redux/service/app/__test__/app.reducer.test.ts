import { appReducer } from '../app.reducer';
import { changeTempUnit } from '../app.action';

describe('App Reducer', () => {
  it('returns the initial state', () => {
    const action = {
      type: 'DUMMY_ACTION',
    };
    const reducer = appReducer(undefined, action);

    expect(reducer).toMatchSnapshot();
  });

  describe('handles the changeTempUnit action', () => {
    it('returns tempUnit with given payload is true', () => {
      const changeTempUnitAction = changeTempUnit(true);
      const reducer = appReducer(undefined, changeTempUnitAction);

      expect(reducer).toMatchSnapshot();
    });

    it('returns tempUnit with given payload is false', () => {
      const changeTempUnitAction = changeTempUnit(false);
      const reducer = appReducer(undefined, changeTempUnitAction);

      expect(reducer).toMatchSnapshot();
    });
  });
});
