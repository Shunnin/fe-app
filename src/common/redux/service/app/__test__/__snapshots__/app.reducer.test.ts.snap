// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Reducer handles the changeTempUnit action returns tempUnit with given payload is false 1`] = `
Object {
  "tempUnit": "FAHRENHEIT",
}
`;

exports[`App Reducer handles the changeTempUnit action returns tempUnit with given payload is true 1`] = `
Object {
  "tempUnit": "CELSIUS",
}
`;

exports[`App Reducer returns the initial state 1`] = `
Object {
  "tempUnit": "CELSIUS",
}
`;
